### ISSUE TITLE
A better way for customizing built-in models and models from components

### ISSUE THREAD

**Issue by bajtos** [strongloop/loopback#397]
Consider an application with a "User" model that has few extra properties or methods compared to the built-in User model.

At the moment, the recommended solution is to create a `user` model extending the built-in `User` model, which is rather silly.

**What we would like to have instead**

Support `redefine:true` as an alternative to `base: "User"`. A redefined "User" model is defined in `common/models/user.json` and scaffolded by `yo loopback`. This will allow redefinition of common models in the server facet.

In the long term (loopback 3.x), we were discussing a mixin-based approach, where User would be just a thin wrapper mixing in login/logout and other features provided by the current User implementation.

---

**Original issue description**

Possible solutions:
1. Implement namespaces, move all built-in models to `loopback` namespace. Application models can live in the global namespace, i.e. `User` extends `loopback.User`.
2. Implement means allowing applications to modify the built-in models, so that they can use them directly.

See also https://github.com/strongloop/loopback-datasource-juggler/issues/119

A probably related problem: the canonical project layout has three facets - "common", "server", "client". In many cases it is needed to provide different implementation of model methods on the server. The current approach is to create a new "server" model extending the shared "common" model and make any changes in this subclass. While this keeps the design clean, it's also rather cumbersome.

**Comment by bajtos** [comment]
Related: how to customize the `Change` model used for change-tracking. https://github.com/strongloop/loopback/issues/926

**Comment by uNki23** [comment]
hi there! 

i'm facing the same problems: pain to customize the built in User model. Extending it by using my own 'user' model isn't working good with angular (injecting 'user' into the function doesn't work, 'User' does, but refers to the base model - see: https://stackoverflow.com/questions/29458355/loopback-angularjs-extending-user-model?answertab=active#tab-top) is there any best practice solution to customize the base User model? i'm doing in the boot-script way with code.

**Comment by pulkitsinghal** [comment]
@uNki23 - I switched to the nomenclature where I create a file named `user-model.json` which extends the base `User` and named it `UserModel` ... it prettifies everything for me in core and angular code ... a bit wordy though.

**Comment by alFReD-NSH** [comment]
I want to add a property to the built-in Role model. Any idea how to do this currently?

**Comment by uNki23** [comment]
correct me if i'm wrong, but it seems to work fine if you just "use" the property.

here is a sample function within an AngularJS UserController:

vm.addUser = function () {
            User.create({
                firstName: vm.firstName,
                lastName: vm.lastName,
                division: vm.division,
                status: vm.status,
                email: vm.email,
                password: vm.password
            })
                .$promise
                .then(function (user) {
                    console.log('Added User: ' + user.firstName + ' ' + user.lastName + '.');
                });
        };

works fine to me, although the first four properties are no properties of the base User model. I've not extended the built in User model in any way. the data is written to mongodb without any problem and i can access the properties by using User.firstName or User.division e.g.

**Comment by alFReD-NSH** [comment]
I'm trying to make it work in postgresql. It would work, if I built it in the schema myself but then I want loopback build the schemas.

**Comment by bajtos** [comment]
> I want to add a property to the built-in Role model. Any idea how to do this currently?

You should be able to do it using the following workaround:

```js
// server/server.js

var loopback = require('loopback');
// etc.

// ADD THIS LINE BEFORE CALLING boot()
loopback.Role.defineProperty('name', { type: 'string' });

boot(app, __dirname);
// etc.
```

**Comment by mikesparr** [comment]
If anyone needs example of how to move the built-in user models to a datasource, I used the `example-mysql` repo exercise and added a few more `automigrate` functions to the script and changed DS for the built-in models.  https://github.com/mikesparr/loopback-example-users-to-mysql  Hope this is helpful for others.

**Comment by clarkbw** [comment]
Didn't notice this discussion previously, I made an initial pass at creating a User model purely from mixins in #1578.  I agree with the above that defining everything in JS instead of the JSON does seem like a lot of overhead, if a mixin could provide a JSON definition file to be merged that might be an interesting way to clean up some of the code.  You can see in my I created a 'properties' directory ([email propery](https://github.com/clarkbw/loopback/commit/c82467bfe1b05c1c68f4d773540bdada5f7992c7#diff-413a89ae6430ef10b602b21ba7a1402dR7)) because I assumed I'd be reusing those properties in higher level mixins

**Comment by mbeauchamp7** [comment]
Hi, has this been resolved?

**Comment by DragonTorus** [comment]
Hi! Is there any oficial, normal way(not workaround) to redefine base Model and its relations in loopback 2.27.0 & jugler 2.45.2? Thanks.

**Comment by JonnyBGod** [comment]
How is this feature coming up? Any timeframe for this to be implemented?

The lack of this feature is a big burden especially when you are crating a publicly available API where model names should make sense. Client is very different from User...

**Comment by jcolemorrison** [comment]
I was content to just leave it as `user` for now, despite the fact it's pretty goofy and causes confusion.  But when just setting up something to autoupdate the base User model and the extended `user` model, both seem to automigrate each other.  Therefore just wound up having to switch the `user` over to `Client`.

And @JonnyBGod is on point.  Above anything else it creates a ridiculous goof in terms of naming and public facing APIs.  As silly as it sounds, this one limitation is a constant pain point I tend to be annoyed with more than most other things in loopback.

**Comment by MichaelPlaxico** [comment]
This is still a relevant issue that causes serious problems in terms of API design and integration with third-party libraries. Is a fix being planned?

**Comment by bajtos** [comment]
> This is still a relevant issue that causes serious problems in terms of API design and integration with third-party libraries. Is a fix being planned?

Not in the current code base. We are working on a new version of LoopBack -
 see https://github.com/strongloop/loopback-next - where issues like this one should be easier to address.

**Comment by RishiMahendru** [comment]
Hi @bajtos is there a way to override persistedModel class methods and add our custom logic to used across all custom models

**Comment by bajtos** [comment]
@RishiMahendru sure. Create your own base model inheriting from PersistedModel, and then use this model as a base for all your application models.

This question is pretty off-topic here, please post in on StackOverflow if you need more information.
